snuba/datasets/events.py

```diff
@@ -27,7 +26,7 @@
 NESTED_COL_EXPR_RE = re.compile(r'^(tags|contexts)\[([a-zA-Z0-9_\.:-]+)\]$')
 
 
-class EventsDataset(Dataset):
+class EventsDataset(TimeSeriesDataset):
```

Comment:
I am concerned this can easily become a mixin, it seems to me we are going towards a lot of datasets customization that each dataset will pick features from different places. (think about a CDC dataset with timeseries support, with delegating event processor, and tag promotion).
Would it make sense to make TimeSeriesDataset either a mixin (then we would need to define the semantics of column_expr when two mixins define it)
Or, better in my opinion, go for composition and make TimeSeriesSupport a class that extends "ColumnExpression" that exposes the column_expr method. The column_expr method on the dataset would delegate the column processing to the ColumnExpression classes defined in the dataset config ?

Reply:
> I am concerned this can easily become a mixin, it seems to me we are going towards a lot of datasets customization that each dataset will pick features from different places. (think about a CDC dataset with timeseries support, with delegating event processor, and tag promotion).

I share this opinion.

> Would it make sense to make TimeSeriesDataset either a mixin (then we would need to define the semantics of column_expr when two mixins define it)

I'd like to avoid mixins (and diamond inheritance) for these reasons as well.

> Or, better in my opinion, go for composition and make TimeSeriesSupport a class that extends "ColumnExpression" that exposes the column_expr method. The column_expr method on the dataset would delegate the column processing to the ColumnExpression classes defined in the dataset config ?

I think that I agree this should be the longer-term approach that we take here, but I'm a little hesitant to make that broad of a change until it's necessary and/or we have a better idea about what other kinds of query-time functionality needs to be encapsulated in the time-series dataset. (I am a little concerned that using several different column expression extensions on the same dataset will lead to the same issues as mixins or multiple inheritance, but at least in that case the scope is narrower.)

In other words — I'm pretty confident we'll end up doing this (or something pretty similar), I just don't know as we *need* to right now. Maybe a compromise would be to make a ticket to follow up on this to make sure that we address it (even if not absolutely necessary) if/when we have some time to circle back and make general improvements; otherwise we'll just tackle it if/when it's a necessity?

Reply:
sure, mind filing the ticket ?

Reply:
This is recorded at SNS-250 now.